tracker/event_tracker_store.go

```diff
@@ -0,0 +1,331 @@
+func (e *Entry) StoreLogs(logs []*ethgo.Log) error {
+	if len(logs) == 0 {
+		return nil
+	}
+
+	var lastBlockNumber uint64
+
+	if err := e.conn.Update(func(tx *bolt.Tx) error {
+		bucketLogs := tx.Bucket(e.bucketLogs)
+		lastLogIndx := getLastIndex(bucketLogs)
+
+		for idx, log := range logs {
+			logIdx := lastLogIndx + uint64(idx)
+
+			val, err := log.MarshalJSON()
+			if err != nil {
+				return err
+			}
+
+			if err := bucketLogs.Put(itob(logIdx), val); err != nil {
+				return err
+			}
+		}
+
+		e.logger.Info("write event logs",
+			"from", lastLogIndx, "to", lastLogIndx+uint64(len(logs))-1, "block", lastBlockNumber)
+
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	lastBlockNumber = logs[len(logs)-1].BlockNumber
+
+	notifyLogs, lastProcessedIdx, err := e.getFinalizedLogs(lastBlockNumber - e.finalityDepth)
+	if err != nil {
+		return err
+	}
+
+	return e.notifyFinalizedLogs(notifyLogs, lastProcessedIdx)
+}
+
+// RemoveLogs implements the store interface
+func (e *Entry) RemoveLogs(indx uint64) error {
+	return e.conn.Update(func(tx *bolt.Tx) error {
+		cursorLogs := tx.Bucket(e.bucketLogs).Cursor()
+
+		// remove logs
+		for k, _ := cursorLogs.Seek(itob(indx)); k != nil; k, _ = cursorLogs.Next() {
+			if err := cursorLogs.Delete(); err != nil {
+				return err
+			}
+		}
+
+		return nil
+	})
+}
+
+// GetLog implements the store interface
+func (e *Entry) GetLog(idx uint64, log *ethgo.Log) error {
+	return e.conn.View(func(tx *bolt.Tx) error {
+		val := tx.Bucket(e.bucketLogs).Get(itob(idx))
+		if val == nil {
+			return fmt.Errorf("log not found: %d", idx)
+		}
+
+		return log.UnmarshalJSON(val)
+	})
+}
+
+func (e *Entry) getFinalizedLogs(untilBlockNumber uint64) ([]*ethgo.Log, uint64, error) {
+	var (
+		logs             []*ethgo.Log
+		lastProcessedIdx uint64 = 0
+	)
+
+	if err := e.conn.View(func(tx *bolt.Tx) error {
+		bucketLastProcessedBlock := tx.Bucket(e.bucketNextToProcess)
+		cursorLogs := tx.Bucket(e.bucketLogs).Cursor()
+
+		var key, value []byte
+
+		// pick first unprocessed block
+		if _, val := bucketLastProcessedBlock.Cursor().First(); val != nil {
+			key, value = cursorLogs.Seek(val)
+		} else {
+			key, value = cursorLogs.First()
+		}
+
+		for ; key != nil; key, value = cursorLogs.Next() {
+			log := &ethgo.Log{}
+			if err := json.Unmarshal(value, log); err != nil {
+				return err
+			}
+
+			if log.BlockNumber > untilBlockNumber {
+				break
+			}
+
+			logs = append(logs, log)
+			lastProcessedIdx = btoi(key)
+		}
+
+		return nil
+	}); err != nil {
+		return nil, 0, err
+	}
+
+	return logs, lastProcessedIdx, nil
+}
+
+func (e *Entry) notifyFinalizedLogs(logs []*ethgo.Log, lastProcessedIdx uint64) error {
+	if len(logs) == 0 {
+		return nil
+	}
+
+	e.logger.Info("notify event logs", "len", len(logs), "last processed", lastProcessedIdx)
+
+	if err := e.conn.Update(func(tx *bolt.Tx) error {
+		return tx.Bucket(e.bucketNextToProcess).Put([]byte("0"), itob(lastProcessedIdx+1))
```

Comment:
Could you extract this singular key (`[]byte("0")`) to a variable?

Reply:
fixed https://github.com/0xPolygon/polygon-edge/pull/1228/commits/7c00602ddd77823dd9574710897a8307d0399e5f